src/core/public/overlays/flyout.tsx

```diff
@@ -0,0 +1,111 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/* tslint:disable max-classes-per-file */
+
+import { EuiFlyout } from '@elastic/eui';
+import React from 'react';
+import { render, unmountComponentAtNode } from 'react-dom';
+import { Observable, Subject } from 'rxjs';
+import { I18nSetup } from '../i18n';
+
+/**
+ * A FlyoutSession describes the session of one opened flyout panel. It offers
+ * methods to close the flyout panel again. If you open a flyout panel you should make
+ * sure you call {@link FlyoutRef#close} when it should be closed.
+ * Since a flyout could also be closed without calling this method (e.g. because
+ * the user closes it), you must listen to the "closed" event on this instance.
+ * It will be emitted whenever the flyout will be closed and you should throw
+ * away your reference to this instance whenever you receive that event.
+ *
+ * @public
+ */
+export class FlyoutRef {
+  /**
+   * An Observable that will emit and complete once this flyout is closed,
+   * by the user or by closing it from the outside via valling {@link #close}.
+   */
+  public readonly onClose$: Observable<void>;
```

Comment:
I am wondering (and I think that might be caused that I missed one push on my previous PR), if this shouldn't rather be a Promise instead? Every flyout can only be closed once, so I don't see there is a good use-case for having an observable here. I just change that to being a promise and instead use `this.closeSubject.toPromise()` in the constructor below. This also has the nicer side effect, that we can use `await/async` syntax on this more easily.

Reply:
I just saw I actually didn't miss any push, you just reverted it in https://github.com/elastic/kibana/pull/34261/commits/c661a51c2e14f95ec96755d0c7c261e37af51f59. Could you elaborate on the reasons for using an observable here for something we know can only be resolved once?

Reply:
This is partly convention, we want plugins to expose all data as observables to more accurately model that it can constantly change and make our code more reactive.

I also think observables are more than simply a stream of future values, it's more than a promise that can resolve multiple times. Observables provide a common abstraction over synchronous/asynchronous, changing and static data that allows for easily combining and operating over different types of data in powerful ways (especially with the large operator library).

I can't think of a good example, but something like... `userInput$.pipe(takeUntil(flyoutRef.onClose$), switchMap...)`

On the other hand it's really easy to convert between Observables and Promises, so just as you could go toPromise on an observable, you could use fromPromise.

Perhaps exposing an observable will encourage others to reach for RxJS operators instead of immediately reaching for procedural `await`-based code.

I would definitely love to hear other people's input on this.

Reply:
While I agree with the general benefits of having the rich async layer that you can use, I personally not a huge fan using it in places where things can only resolve exactly once. As you mentioned it's easy enough converting to an observable from a promise when you need it, but using a promise, but using an Observable, the whole API is build in a way, that it CAN actually resolve multiple times. So instead of having the technical API represent, that something can only resolve a single time, we use an API that technically can "resolve multiple times", and just put that into documentation: "Well this can only resolve once actually". I prefer more to express the actual working as good as we can via the (typed) API that your code editor will complete on, etc.

But as you mentioned I would really welcome some more opinions in this discussion :-)

Reply:
I think it's reasonable to expose a promise when you definitely know something can only be resolved a single time. We lose some inherent flexibility with observables when we do this, but we also gain some clarity in terms of how we expect the thing to be used. We've already made this tradeoff before with plugin and service lifecycle functions, which return promises rather than observables.

That said, in this particular case it isn't clear to me what's the appropriate behavior. With things like the siem timeline, there's now a use case for a flyout that "follows" you around Kibana and can be expanded/collapsed at will. In that case, I could potentially see `close` being a recurring operation that would be more appropriate for an observable than a promise. I can also imagine designs that wouldn't couple that behavior to the flyout service at all.

Reply:
In case you "reopen" a flyout, you would basically get a new FlyoutRef, every FlyoutRef can only have one `onClose` called, then it's done. Opening another flyout (even if it's the same content) will lead to a new FlyoutRef.

Reply:
I agree that a promise-based API makes it more explicit from the API signature that only one value will be emitted/resolved.

So the tradeoff is then between:
1. A Promise-based API which makes it more explicit that a developer should expect only one return value.
2. An Observable-based API which has a richer more powerful abstraction that might encourage more declarative code and a better architecture.

In this case none of the other benefits of Observables like lazy evaluation and the ability to cancel is immediately obvious. The architecture argument is speculative and it will require much more than exposing observables to get teams to write more declarative code. So I'm happy to go with (1).